Refactor bsblan coordinator

liudger opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update coordinator to use models

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

liudger committed:
chore: Refactor BSBLanUpdateCoordinator to improve code readability and maintainability

liudger committed:
feat: Add BSBLan integration models

This commit adds the models for the BSB-Lan integration. It includes a dataclass for the BSBLanCoordinatorData, which stores the state and sensor information.

liudger committed:
refactor: Update BSBLANClimate class to use DataUpdateCoordinator without specifying the State type

liudger committed:
chore: Remove unused Sensor import in BSBLan models

liudger commented:
@joostlek Is this the size that is right?

joostlek commented on homeassistant/components/bsblan/climate.py:
```diff
@@ -67,10 +65,12 @@ async def async_setup_entry(
 
 class BSBLANClimate(
-    BSBLANEntity, CoordinatorEntity[DataUpdateCoordinator[State]], ClimateEntity
+    BSBLANEntity, CoordinatorEntity["BSBLanUpdateCoordinator"], ClimateEntity
```
So what I'm wondering, why does it inherit both the BSBLANentity and the coordinator entity, while the BSBLANentity already inherits the coordinator one

liudger replied:
lol indeed this makes no sense. I need to update the entity too to inherit the coordinator. Cutting the pull request in pieces is quite difficult

joostlek replied:
I am fine with making sure that the entity inherits the coordinator in this PR

joostlek commented on homeassistant/components/bsblan/climate.py:
```diff
@@ -67,10 +65,12 @@ async def async_setup_entry(
 ):
     """Defines a BSBLAN climate device."""
 
+    coordinator: BSBLanUpdateCoordinator
```
This would not be needed

joostlek commented on homeassistant/components/bsblan/coordinator.py:
```diff
@@ -1,57 +1,57 @@
         """Initialize the BSB-Lan coordinator."""
-
         self.client = client
+        self.bsblan_config_entry = config_entry
```
Why do we do this? It's already accessible via self.config_entry. before the type was overwritten, making sure you don't have type issues

joostlek commented on homeassistant/components/bsblan/models.py:
```diff
@@ -0,0 +1,12 @@
+
+
+@dataclass
+class BSBLanCoordinatorData:
```
Why do we store this here instead of in the coordinator?

liudger replied:
Why not? It looks clean to store all data classes in models

joostlek replied:
Because it's the data of the coordinator so it looks more logical to have it in the coordinator

liudger committed:
feat: Refactor BSBLanEntity to use CoordinatorEntity

The BSBLanEntity class has been refactored to inherit from the CoordinatorEntity class, which provides better integration with the update coordinator. This change improves code readability and maintainability.

liudger committed:
refactor: Remove unused config_entry variable in BSBLanUpdateCoordinator

liudger committed:
refactor: Update BSBLANClimate class to use DataUpdateCoordinator

Refactor the BSBLANClimate class to use the Coordinator of the entity

liudger committed:
refactor: Update tests to use the new structure

liudger commented:
OK I made quite some mistakes. Good to make smaller pr
Do I need to move the data class from the coordinator to the coordinator itself?

liudger commented:
accidentally clicked on re-review

liudger committed:
fix coverage

 it should be the same as before

liudger committed:
refactor: moved dataclass BSBLanCoordinatorData

liudger commented:
Moved data class coordinator. Not sure why codecov failed

joostlek commented:
codecov is flakey

joostlek commented on homeassistant/components/bsblan/entity.py:
```diff
@@ -1,41 +1,39 @@
+            configuration_url=f"http://{host}",
         )
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self.coordinator.last_update_success
```
This is the default

joostlek commented on homeassistant/components/bsblan/models.py:
```diff
@@ -0,0 +1,20 @@
+    from .coordinator import BSBLanUpdateCoordinator
+
+
+@dataclass
+class BSBLanData:
+    """BSBLan data stored in the entity."""
+
+    coordinator: "BSBLanUpdateCoordinator"
+    client: BSBLAN
+    device: Device
+    info: Info
+    static: StaticState
```
I personally think this is nicer to live in `__init__.py`

joostlek commented on tests/components/bsblan/snapshots/test_diagnostics.ambr:
```diff
@@ -1,12 +1,68 @@
     }),
+    'entry': dict({
+      'data': dict({
+        'host': '127.0.0.1',
+        'passkey': '1234',
+        'password': 'admin1234',
```
Should we redact this?

liudger replied:
I thought homeassistant automatically redacts this info? Should I add an redact function?

joostlek replied:
You have to redact the data yourself, there's a function for that

liudger commented on homeassistant/components/bsblan/entity.py:
```diff
@@ -1,41 +1,39 @@
+        """Initialize BSBLan entity."""
+        super().__init__(coordinator)
+        self._data = data
+        host = (
+            coordinator.config_entry.data["host"]
+            if coordinator.config_entry
+            else "unknown"
+        )
```
Even when I overwrite I get this error.
```homeassistant/components/bsblan/entity.py:22: error: Item "None" of "ConfigEntry[Any] | None" has no attribute "data"  [union-attr]```

```
def __init__(
        self,
        hass: HomeAssistant,
        config_entry: ConfigEntry,
        client: BSBLAN,
    ) -> None:
        """Initialize the BSB-Lan coordinator."""
        super().__init__(
            hass,
            logger=LOGGER,
            name=f"{DOMAIN}_{config_entry.data[CONF_HOST]}",
            update_interval=self._get_update_interval(),
        )
        self.client = client
        self.config_entry = config_entry
 ```
 
 am I doing something wrong?

joostlek replied:
Outside of the constructor 

```
config_entry: ConfigEntry
```

liudger committed:
use the data class inside init

liudger committed:
refactor: Remove unused config_entry variable in BSBLanUpdateCoordinator

liudger committed:
refactor: use BSBLanData from init

liudger committed:
remove entry data from diagnostics

joostlek commented on homeassistant/components/bsblan/coordinator.py:
```diff
@@ -28,30 +33,31 @@ def __init__(
 
-    async def _async_update_data(self) -> State:
-        """Get state from BSB-Lan device."""
+    def _get_update_interval(self) -> timedelta:
```
This is only run once btw, so it doesn't change like before

joostlek commented on homeassistant/components/bsblan/diagnostics.py:
```diff
@@ -7,17 +7,23 @@
         "info": data.info.to_dict(),
         "device": data.device.to_dict(),
-        "state": data.coordinator.data.to_dict(),
+        "coordinator_data": {
+            "state": data.coordinator.data.state.to_dict()
+            if data.coordinator.data
+            else {},
+        },
+        "static": data.static.to_dict(),
```
Ideally we change this in a follow up btw, let's do that as its not redacted yet

liudger replied:
So we leave this for now and do a merge?

liudger committed:
fix: add random interval back

liudger committed:
refactor: Simplify coordinator_data assignment in async_get_config_entry_diagnostics

joostlek commented:
Sent a message on discord :)

liudger committed:
revert back to original except dataclass import

joostlek commented on homeassistant/components/bsblan/entity.py:
```diff
@@ -1,41 +1,29 @@
+        super().__init__(coordinator, data)
+        host = self.coordinator.config_entry.data["host"]
         self._attr_device_info = DeviceInfo(
-            connections={(CONNECTION_NETWORK_MAC, format_mac(device.MAC))},
```
Let's not remove this

liudger committed:
revert: Add MAC address back to device info in BSBLanEntity
